RPIPs/RPIP-onchain-pdao.md

```diff
@@ -0,0 +1,457 @@
+percent of $P$. And it's veto quorum is set to `proposal.veto.quorum` percent of $P$.
+
+If a proposal is challenged, the proposer MAY respond to the challenge by providing a new pollard where the root node
+is the challenged index and a merkle proof from the challenged index back to the proposal root. A proposer will be
+unable to respond if the response requires submitting a pollard containing leaf nodes where the leaf nodes do not match
+the actual values on chain. In this way, they are forced to not respond for `proposal.challenge.period` time. After
+which, the proposal can be defeated.
+
+If a proposal is defeated, the proposer forfeits their bond which is divided proportionally amongst the challengers who
+contributed to the proposal's defeat.
+
+To raise a proposal, a node operator MUST
+call `propose(string proposalMessage, bytes _payload, uint32 blockNumber, Type.Node[] treeNodes)`
+on `RocketDAOProtocolProposals`. Where `proposalMessage` is an arbitrary string, `_payload` is an ABI encoded function
+call to `RocketDAOProtocolProposals`, `blockNumber` is the block number used to generate the merkle pollard,
+and `treeNodes` is the merkle pollard.
+
+To respond to a challenge, the proposer MUST
+call `submitRoot(uint256 proposalID, uint256 index, Types.Node[] witness, Types.Node[] nodes`
+on `RocketDAOProtocolVerifier`. Where `proposalID` is the proposal's ID, `index` is the challenged index, `witness` is
+the merkle proof and `nodes` is the new merkle pollard from the challenged index.
+
+To claim a refund for a proposal bond and rewards from any invalid challenges, the proposer MUST
+call `claimBondProposer(uint256 proposalID, uint256[] indices)` on `RocketDAOProtocolVerifier`. Where `proposalID` is
+the proposal's ID and `indices` is an array of indices that are being claimed. The index "1" refers to the proposal
+root. Claiming this index will return the proposal bond.
+
+### Verifying
+
+Any node MAY challenge a proposal by supplying an index into the merkle-sum tree that they are alleging is incorrect.
+The challenger MUST have at least `proposal.challenge.bond` unlocked effective RPL stake. And this amount of the
+challenger's RPL SHALL BE locked until the proposal is defeated.
+
+The index which is alleged to be incorrect MUST be an index of a node in one of the pollards that the proposer has
+submitted. Either the initial pollard submitted with the proposal, or any subsequent response pollards to a challenge.
+
+If the initial pollard submitted by the proposer is invalid, they will be unable to respond to challenges which require
+the proposer to provide information that is known on chain. It will take a few rounds of challenges and responses to
+reach the last layer of the merkle-sum tree. After which, the proposer will no longer be able to respond to challenges.
+
+After `proposal.challenge.period` has passed and the proposer has not responded to the challenge, any one MAY
```

Comment:
Is challenge period refreshed on each challenge? Otherwise what prevents a challenge one block before the end of the period?

And the flip side: if it _is_ refreshed, that means the challenger can delay the vote by the refresh time times the number of challenge rounds needed. How long is that and is it acceptable?

Reply:
Each challenge has it's own period. Once I challenge, the proposer has `proposal.challenge.period` to respond. Once they respond, I can submit a new challenge.

Challenging does not reset the proposal waiting period. The waiting period for a proposal needs to be long enough for a full challenge/response process to take place. 

The waiting period is 1 week, with a 1 hour challenge period, that provides plenty of time for verifiers to perform the fraud proof process. The number of challenge rounds required depends on the node operator count. But it grows at a `log2(N^2)/5` rate. With the current node operator count (3200), 5 rounds of challenges are required. At 10k node operators, it requires 6 rounds. 

Challenges of different paths can happen in parallel too.